### PR TITLE
[WIP] Bounding box

### PR DESCRIPTION
**Pull request by creviera**
**boundingBox** queries the db for all moving animals and changes the direction of an animal so that the animal always remains in the bounding box.

I'd like to figure out if there is a better way to use the room's subscribe constraint solver to only call `bounce_off_wall` when an animal is at a position that intersects with the bounding box. This might not be possible given the constraints of the constraint solver. @jedahan do you have any thoughts about how this could be improved?

There is also a bug where an animal begins to slowly escape the bounding box - the speed gets faster and faster the closer an animal nears escape but the direction continues to change back and forth. It might be a problem with the way I'm not updating an animals position and only their speed if they intersect with a wall. It _might_ help to move the animal's position but I'm not sure how to update that yet.

### CODE REVIEW

**Comment by jedahan** [comment]
So the issue with pumba escaping has to do with `processes/move` updating location whenever speed changes. I think its an off-by-one error in that processes/move can write multiple updates per frame. I'm gonna test out some fixes and once I feel good about them, I'll merge this.

As far as adding smarter queries to the constraint solver, I wanna see how far we can go with a limited language first. If you want to propose a syntax of how you would like to write this code in an ideal world, that would be super helpful though to see a possible future.

File: src/processes/boundingBox.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+module.exports = room => {
+  if (!room) {
+    const Room = require('@living-room/client-js')
+    room = new Room()
+  }
+
+  const box = [0.25, 0.25, 0.5, 0.5]
```
Comment (jedahan) [inline src/processes/boundingBox.js]: Changing this to 
```js
const box = { x: 0.25, y: 0.25, w: 0.5, h: 0.5 }
```
I think will make it easier to read the assert code below

Comment (jedahan) [reply src/processes/boundingBox.js]: In addition, maybe doing a subscription to change the box from within the room, like
```js
room.subscribe(`set the animal bounding box to $x, $y, $w, $h`, newBox => box = newBox)
```

File: src/processes/boundingBox.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  room.assert([
+    `draw a (255, 255, 255) line from (${box[0]}, ${box[1]}) to (${box[0] + box[2]}, ${box[1]})`,
+    `draw a (255, 255, 255) line from (${box[0] + box[2]}, ${box[1]}) to (${box[0] + box[2]}, ${box[1] + box[3]})`,
+    `draw a (255, 255, 255) line from (${box[0] + box[2]}, ${box[1] + box[3]}) to (${box[0]}, ${box[1] + box[3]})`,
+    `draw a (255, 255, 255) line from (${box[0]}, ${box[1] + box[3]}) to (${box[0]}, ${box[1]})`
+  ])
+
+  const bounce_off_wall = async function (changes) {
```
Comment (jedahan) [inline src/processes/boundingBox.js]: we can remove async now!

File: src/processes/boundingBox.js (hunk 1 above)
Comment (jedahan) [reply src/processes/boundingBox.js]: Also how do you feel about defaults like 0, 0, 1, 1 so we can use the whole table?

File: src/processes/boundingBox.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    for (animal of changes.assertions) {
+      if (
+        animal.x <= box[0] ||
+        animal.y <= box[1] ||
+        animal.x >= box[2] ||
+        animal.y >= box[3]
+      ) {
+        room.retract(`${animal.name} has speed (${animal.dx}, ${animal.dy})`)
+        room.assert(`${animal.name} has speed (${-animal.dx}, ${-animal.dy})`)
```
Comment (jedahan) [inline src/processes/boundingBox.js]: So this always reverses both directions, but I'd prefer if we fixed out if we need to reverse just one direction, e.g.
```js
const dx = (animal.x <= box[0] || animal.x >= box[2]) ? -animal.dx : animal.dx
const dy = (animal.y <= box[1] || animal.y >= box[3]) ? - animal.dy : animal.dy
if (dx != animal.dx || dy != animal.dy) {
  //...retractions/assertions here
}
```

**Comment by allieoop** [comment]
@jedahan - when I isolated and tested your change to `move.js` on master, this command stopped working: `pumba has speed (0.008, 0.008)`

File: src/processes/boundingBox.js (hunk 3 above)
Comment (allieoop) [reply src/processes/boundingBox.js]: Good call

**Comment by jedahan** [comment]
Okay I'll take a look and see whats the issue in a couple of hours

**Comment by jedahan** [comment]
Okay I fixed move.js and rebased this branch on master, can you test and see that it works? it looks great from my end.

**Comment by allieoop** [comment]
@jedahan - I moved the draw code into `boundingBoxIlluminator.js` which is making it easier to debug the issue with the strange animal movement behavior when inside `bounceOffWall`:

```node
  // If an animal intersects with a boundingBox wall "bounce" the animal in the
  // opposite direction
  const bounceOffWall = function (changes) {
    for (i of changes.assertions) {
      if (i.x <= i.x0 || i.x >= i.w || i.y <= i.y0 || i.y >= i.h) {
        room.retract(`${i.name} has speed (${i.dx}, ${i.dy})`)
        room.assert(`${i.name} has speed (${-i.dx}, ${-i.dy})`)
      }
    }
  }
```
I'm not sure why `(i.x <= i.x0 || i.x >= i.w || i.y <= i.y0 || i.y >= i.h)` is being satisfied when an animal at (x, y) isn't intersecting with a wall at (x0, y0) of size w x h. Also, looking at this now is making me realize that logic needs to be added for the case where a moving animal starts outside a bounding box. Anyway, thought I'd bring this to your attention now that it's easier to see the issue with the current code and in case you think there could still be a bug in the `move.js` code.

**Comment by jedahan** [comment]
I'll take a look at this tonight, and see if theres anything obvious about the movement. I full expect animals to move outside the bounding box if their movement speed is high enough, but (0.01, 0.01) should be sane

**Comment by jedahan** [comment]
Err, gonna look at this over the weekend... had some trouble updating service-js and client-js which made it hard for me to test.

**Comment by jedahan** [comment]
Okay I can describe whats happening with the jumping to other side of the table and moving down.

I added a feature in display.js that seems to be a bug - if a coordinate is between -1 and 1, the display interprets it as normalized, otherwise its considered absolute. So like, when an animal goes from `(whatever, 0.9)` to `(whatever, 1.1)` it jumps to the other side and starts slowly creeping down.

Instead of checking the range, I can check if the number is floating point or integer.

The boundaries thing I will investigate shortly.

**Pull request closed without merging**